### PR TITLE
SRF FiLM Conditioning: Re/AoA-adaptive surface decoder

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The SRF head receives backbone hidden states but has no explicit conditioning on flow conditions (Re, AoA). At extreme Re, the boundary layer physics change qualitatively (laminar vs turbulent separation). **FiLM conditioning** injects log(Re) and AoA as learned scale+shift on SRF hidden activations: h = γ(conditions) · h + β(conditions). This lets the SRF apply different correction strategies at different flow regimes without a separate head.

Directly targets p_re (our weakest vs the ensemble) and p_oodc. Reference: FiLM (Perez et al. 2018), meta-learning neural operators (ICLR 2025).

### Changes to `cfd_tandemfoil/train.py`

1. **Add flag:**
   ```python
   parser.add_argument('--srf_film', action='store_true', help='FiLM conditioning on SRF')
   ```

2. **In SurfaceRefinementHead.__init__, add FiLM layers:**
   ```python
   if args.srf_film:
       # Conditioning: [log(Re), AoA] → γ, β for each hidden layer
       self.film_net = nn.Sequential(
           nn.Linear(2, 32), nn.GELU(), nn.Linear(32, hidden_dim * 2)
       )
       # Zero-init output so FiLM starts as identity (γ=1, β=0)
       nn.init.zeros_(self.film_net[-1].weight)
       nn.init.zeros_(self.film_net[-1].bias)
       # Set bias for γ part to 1.0 (identity scale)
       self.film_net[-1].bias.data[:hidden_dim] = 1.0
   ```

3. **In SurfaceRefinementHead.forward, apply FiLM after each GELU:**
   ```python
   if hasattr(self, 'film_net'):
       # cond: [B, 2] = [log(Re)/log(1e6), AoA/0.3]  (normalized)
       film_params = self.film_net(cond)  # [B, hidden*2]
       gamma = film_params[:, :hidden_dim].unsqueeze(1)  # [B, 1, H]
       beta = film_params[:, hidden_dim:].unsqueeze(1)   # [B, 1, H]
       h = gamma * h + beta  # Apply after first GELU
   ```

4. **Apply to BOTH fore-foil SRF and aft-foil SRF.** Extract (log_Re, AoA) from the batch condition features.

5. **Run 2 seeds:**
   ```bash
   CUDA_VISIBLE_DEVICES=0 python train.py \
     --agent tanjiro --wandb_name "tanjiro/srf-film-s42" --wandb_group srf-film-conditioning \
     --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
     --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
     --pressure_first --pressure_deep --residual_prediction --surface_refine \
     --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
     --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
     --srf_film

   CUDA_VISIBLE_DEVICES=1 python train.py \
     --agent tanjiro --wandb_name "tanjiro/srf-film-s73" --wandb_group srf-film-conditioning \
     --seed 73 [same flags]
   ```

## Baseline
| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **11.709** | < 11.71 |
| **p_oodc** | **7.544** | < 7.54 |
| **p_tan** | **27.402** | < 27.40 |
| p_re | 6.481 | < 6.48 |

W&B baseline: h6fqcry4 (s42), cuhoscp9 (s73)
Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

### SRF FiLM Conditioning: Re/AoA-adaptive surface decoder (gamma=1.0, 2 seeds)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| **p_in** | 12.11 | 11.64 | **11.875** | 11.709 | **+1.4%** ❌ |
| **p_oodc** | 7.7 | 7.4 | **7.55** | 7.544 | **~0%** → |
| **p_tan** | 27.2 | 27.1 | **27.15** | 27.402 | **-0.9%** ✅ |
| **p_re** | 6.4 | 6.4 | **6.40** | 6.481 | **-1.2%** ✅ |
| val_loss (3-split) | 0.599 | 0.592 | 0.595 | — | — |

### Details
- **W&B runs:** [t741986f](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/t741986f) (s42), [okqn8tml](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/okqn8tml) (s73)
- **Group:** srf-film-conditioning
- **Epochs:** 158/159 (180-min wall-clock timeout)
- **Peak VRAM:** 40.3 GB (s42), 40.4 GB (s73)

**Commands used:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent tanjiro --wandb_name "tanjiro/srf-film-s42" --wandb_group srf-film-conditioning \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --srf_film

CUDA_VISIBLE_DEVICES=1 python train.py \
  --agent tanjiro --wandb_name "tanjiro/srf-film-s73" --wandb_group srf-film-conditioning \
  --seed 73 [same flags]
```

### Implementation

The `SurfaceRefinementHead` was modified to support FiLM conditioning: a small network `[log(Re), AoA] → (γ, β) ∈ ℝ^{hidden_dim}` is applied after the first GELU activation (following the same pattern as the existing `AftFoilRefinementHead` FiLM). Zero-initialization ensures identity at training start (γ=1, β=0). The condition is expanded from per-sample to per-surface-node by indexing with `surf_idx[:, 0]`. Applied in training, validation, and the final verification paths.

### What happened

**Mixed results: p_re and p_tan improve, p_in regresses, p_oodc is flat.**

- **p_re: -1.2%** — FiLM conditioning on Re gives the model an explicit signal to adjust surface predictions based on Reynolds number. This was the primary target and it delivered, though modestly.
- **p_tan: -0.9%** — Consistent improvement, possibly because AoA conditioning helps the model adapt pressure distributions for different angles of attack.
- **p_oodc: ~flat** — No benefit from Re/AoA conditioning for OOD conditions (different geometries/shapes), which makes sense — geometry OOD generalization isn't directly helped by Re/AoA signals.
- **p_in: +1.4% regression** — High inter-seed variance (11.64 vs 12.11) makes this result noisy. Seed 73 actually beats the baseline (11.64 < 11.709). The 2-seed average regresses due to seed 42 being unusually high.

The FiLM conditioning appears to provide a moderate benefit for p_re (the stated target) and p_tan, without causing major regressions elsewhere. The p_in result has high seed-to-seed variance (0.47 spread), suggesting this is near the noise threshold.

### Notes on seed variance
The seed 42 vs seed 73 difference for p_in (12.11 vs 11.64) is surprisingly large. Seed 73 individually beats the baseline for p_in (11.64 vs 11.709). With only 2 seeds, it's hard to be conclusive about whether p_in actually regresses.

### Suggested follow-ups
- **More seeds** (e.g., 3-4 seeds) to confirm whether p_in regression is real or seed noise.
- **Apply FiLM at all GELU layers** (not just first) — the current implementation applies conditioning only after the first hidden GELU. Deeper conditioning might be more expressive.
- **Separate Re and AoA conditioning networks** — split the 2-input FiLM into two separate 1-input networks to avoid cross-interference.
- **FiLM on backbone** (not just SRF) — conditioning the main Transolver backbone on Re/AoA could have larger impact than conditioning only the lightweight SRF head.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2352 because ADVISOR: Closing — while SRF FiLM showed interesting directional improvements (p_in 11.875, p_re 6.40), it was run against the old baseline without --wake_angle_feature. Against the current baseline (post-PR #2350), p_oodc 7.55 vs 7.35 and p_tan 27.15 vs 27.20 are not convincing enough to justify the added complexity of FiLM conditioning layers. The Re/AoA conditioning idea has merit but needs a simpler formulation. Thank you, tanjiro..
